ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/ChainDataPruner.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.chain;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorage;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorageTransaction;
+
+import java.util.Collection;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ChainDataPruner implements BlockAddedObserver {
+  private static final Logger LOG = LoggerFactory.getLogger(ChainDataPruner.class);
+
+  // TODO: cleanup - pruningEnabled will not be required after
+  // https://github.com/hyperledger/besu/pull/4703 is merged.
+  private static boolean pruningEnabled = false;
+  private final BlockchainStorage blockchainStorage;
+  private final ChainDataPrunerStorage prunerStorage;
+  private final long blocksToRetain;
+
+  private final long pruningFrequency;
+
+  public ChainDataPruner(
+      final BlockchainStorage blockchainStorage,
+      final KeyValueStorage storage,
+      final long blocksToRetain,
+      final long pruningFrequency) {
+    this.blockchainStorage = blockchainStorage;
+    this.prunerStorage = new ChainDataPrunerStorage(storage);
+    this.blocksToRetain = blocksToRetain;
+    this.pruningFrequency = pruningFrequency;
+  }
+
+  public static void enablePruning() {
+    pruningEnabled = true;
+  }
+
+  public static boolean isPruningEnabled() {
+    return pruningEnabled;
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    LOG.debug("New block added event: " + event);
+    new Thread(
```

Comment:
This could possibly end up creating an unlimited number of threads. Suggest that we use an ExecutorService instead. The `MonitoredExecutors` could be useful for creating an executor that has metrics.

Reply:
Just tried to use MonitoredExecutors but this introduces a circular dependency between :ethereum:core and :ethereum:eth modules

Reply:
Updated & switched to ThreadPoolExecutor to limit the number of threads.